lib/Analysis/Memory/DefinedMemory.cpp

```diff
@@ -631,6 +643,119 @@ bool DataFlowTraits<ReachDFFwk*>::transferFunction(
   return false;
 }
 
+template<typename FuncT>
+void addLocationToSet(DFRegion *R, const MemoryLocationRange &Loc,
+    const ReachDFFwk &Fwk, FuncT &&Inserter) {
+  typedef MemoryLocationRange::Dimension Dimension;
+  auto LocInfo = Fwk.getDelinearizeInfo()->findRange(Loc.Ptr);
+  auto ArrayPtr = LocInfo.first;
+  if (!ArrayPtr || !ArrayPtr->isDelinearized() || !LocInfo.second->isValid()) {
+    Inserter(Loc);
+    return;
+  }
+  LLVM_DEBUG(
+    for (auto I = 0; I < ArrayPtr->getNumberOfDims(); I++) {
+      dbgs() << "[ARRAY LOCATION] Dimension `" << I << "` size: " <<
+          *ArrayPtr->getDimSize(I) << "\n";
+    }
+  );
+  LLVM_DEBUG(dbgs() << "[ARRAY LOCATION] Array info: " <<
+      *ArrayPtr->getBase() << "\n");
+  LLVM_DEBUG(dbgs() << "[ARRAY LOCATION] IsAddress: " <<
+      ArrayPtr->isAddressOfVariable() << ".\n");
+  auto BaseType = ArrayPtr->getBase()->getType();
+  auto ArrayType = cast<PointerType>(BaseType)->getElementType();
+  if (ArrayPtr->isAddressOfVariable()) {
+    ArrayType = cast<PointerType>(ArrayType)->getElementType();
+    // TODO: ArrayPtr = ?
+  }
+  LLVM_DEBUG(dbgs() << "[ARRAY LOCATION] Array type: " << *ArrayType << ".\n");
+  LLVM_DEBUG(dbgs() << "[ARRAY LOCATION] Element type: " <<
+      ArrayType->getTypeID() << ".\n");
+  auto ArraySizeInfo = arraySize(ArrayType);
+  if (ArraySizeInfo == std::make_tuple(0, 1, ArrayType) &&
+      ArrayPtr->getNumberOfDims() != 1) {
+    LLVM_DEBUG(dbgs() << "[ARRAY LOCATION] Failed to get array size.\n");
+    Inserter(Loc);
+    return;
+  }
+  auto ElementType = std::get<2>(ArraySizeInfo);
+  LLVM_DEBUG(dbgs() << "[ARRAY LOCATION] Array elements type:  " <<
+      *ElementType << ".\n");
+  size_t DimensionN = 0;
+  MemoryLocationRange ResLoc(ArrayPtr->getBase(), 0,
+      uint64_t(Fwk.getDataLayout().getTypeStoreSize(ElementType)));
+  for (auto *S : LocInfo.second->Subscripts) {
+    LLVM_DEBUG(dbgs() << "[ARRAY LOCATION] Subscript: " << *S << "\n");
+    if (!ArrayPtr->isKnownDimSize(DimensionN) && DimensionN != 0) {
+      LLVM_DEBUG(dbgs() << "[ARRAY LOCATION] Failed to get dimension "
+                           "size for required dimension `" << DimensionN <<
+                           "`.\n");
+      break;
+    }
+    Dimension DimInfo;
+    auto DimSize = ArrayPtr->getDimSize(DimensionN);
+    if (DimSize->getSCEVType() == scConstant) {
+      DimInfo.DimSize = *cast<SCEVConstant>(DimSize)->getValue()->
+          getValue().getRawData();
+    } else {
+      LLVM_DEBUG(dbgs() << "[ARRAY LOCATION] Non-constant dimension size.\n");
+      if (DimensionN != 0)
+        break;
+    }
+    auto *SE = Fwk.getScalarEvolution();
+    //LLVM_DEBUG(SE->print(dbgs()); dbgs() << "\n";);
+    auto AddRecInfo = computeSCEVAddRec(S, *SE);
+    if (!AddRecInfo.second)
+      break;
+    auto SCEV = AddRecInfo.first;
+    if (SCEV->getSCEVType() == scConstant) {
+      auto Range = *cast<SCEVConstant>(S)->getValue()->getValue().getRawData();

```

Comment:
А вот индексация по массиву, это знаковое число, `int64_t`.
```suggestion
    if (auto C = dyn_cast<SCEVConstant>(SCEV)) {
      auto Range = C->getAPInt().getSExtValue();
```